util/chunk/iterator.go

```diff
@@ -284,3 +304,165 @@ func (it *iterator4RowPtr) ReachEnd() {
 func (it *iterator4RowPtr) Len() int {
 	return len(it.ptrs)
 }
+
+// Error returns none-nil error if anything wrong happens during the iteration.
+func (it *iterator4RowPtr) Error() error {
+	return nil
+}
+
+// NewIterator4RowContainer create a new iterator for RowContainer
+func NewIterator4RowContainer(c *RowContainer) *iterator4RowContainer {
+	return &iterator4RowContainer{c: c}
+}
+
+type iterator4RowContainer struct {
+	c      *RowContainer
+	chkIdx int
+	rowIdx int
+	err    error
+}
+
+// Len implements the Iterator interface.
+func (it *iterator4RowContainer) Len() int {
+	return it.c.NumRow()
+}
+
+func (it *iterator4RowContainer) setNextPtr() {
+	it.rowIdx++
+	if it.rowIdx == it.c.NumRowsOfChunk(it.chkIdx) {
+		it.rowIdx = 0
+		it.chkIdx++
+	}
+}
+
+// Begin implements the Iterator interface.
+func (it *iterator4RowContainer) Begin() Row {
+	it.chkIdx, it.rowIdx = 0, -1
+	return it.Next()
+}
+
+// Next implements the Iterator interface.
+func (it *iterator4RowContainer) Next() Row {
+	if it.chkIdx >= it.c.NumChunks() {
+		it.ReachEnd()
+		return it.End()
+	}
+	it.setNextPtr()
+	return it.Current()
+}
+
+// Current implements the Iterator interface.
+func (it *iterator4RowContainer) Current() Row {
+	if it.rowIdx < 0 || it.chkIdx >= it.c.NumChunks() {
+		return it.End()
+	}
+	row, err := it.c.GetRow(RowPtr{uint32(it.chkIdx), uint32(it.rowIdx)})
+	if err != nil {
+		it.err = err
+		it.ReachEnd()
+		return it.End()
+	}
+	return row
+}
+
+// End implements the Iterator interface.
+func (it *iterator4RowContainer) End() Row {
+	return Row{}
+}
+
+// ReachEnd implements the Iterator interface.
+func (it *iterator4RowContainer) ReachEnd() {
+	it.chkIdx, it.rowIdx = it.c.NumChunks(), 0
+}
+
+// Error returns none-nil error if anything wrong happens during the iteration.
+func (it *iterator4RowContainer) Error() error {
+	return it.err
+}
+
+// multiIterator joins several iterators together to form a new iterator
+type multiIterator struct {
+	iters   []Iterator
+	numIter int
+	length  int
+	curPtr  int
+	curIter Iterator
+	err     error
+}
+
+// NewMultiIterator creates a new multiIterator
+func NewMultiIterator(iters ...Iterator) Iterator {
+	iter := &multiIterator{}
+	for i := 0; i < len(iters); i++ {
+		if iters[i].Len() > 0 {
+			iter.iters = append(iter.iters, iters[i])
+			iter.length += iters[i].Len()
+		}
+	}
+	iter.numIter = len(iter.iters)
+	return iter
+}
+
+// Len implements the Iterator interface.
+func (it *multiIterator) Len() int {
+	return it.length
+}
+
+// Begin implements the Iterator interface.
+func (it *multiIterator) Begin() Row {
+	it.curPtr = 0
+	if it.numIter > 0 {
+		it.curIter = it.iters[0]
+		it.curIter.Begin()
+	}
+	return it.Current()
+}
+
+// Next implements the Iterator interface.
+func (it *multiIterator) Next() Row {
+	if it.curPtr == it.numIter {
+		return it.End()
+	}
+	next := it.curIter.Next()
+	if next == it.curIter.End() {
```

Comment:
We should do for loop here in case that the next iter is 0-length.

Reply:
hahah, I've removed all iters that have 0 lenght

```
// NewMultiIterator creates a new multiIterator
func NewMultiIterator(iters ...Iterator) Iterator {
	iter := &multiIterator{}
	for i := 0; i < len(iters); i++ {
		if iters[i].Len() > 0 {
			iter.iters = append(iter.iters, iters[i])
			iter.length += iters[i].Len()
		}
	}
	iter.numIter = len(iter.iters)
	return iter
}
```